### DISCUSSION TITLE
CreateRunAsync slow to load when Assistant instructions exceed 9,000 characters

### DISCUSSION THREAD

**Discussion by JBlakeMCAA** [RageAgainstThePixel/com.openai.unity#263] [Q&A]
It takes over 20 seconds to load "Hello" back from the assistant when the instructions are large. CreateRunAsync slow to load when Assistant instructions exceed 9,000 characters
Is is possible to not bring back everything and just the delta? Below is the code I am running the the log output:
  ```csharp
     RunResponse run = await thread.CreateRunAsync(assistant, async streamEvent =>
            {
                switch (streamEvent)
                {
                    case MessageResponse messageEvent:
                        switch (messageEvent.Object)
                        {
                            case "thread.message.delta":
                                Debug.Log($"thread.message.delta: {messageEvent.Object}: \"{messageEvent.Delta.PrintContent()}\"");
                                break;
                        }
                        break;
                }
                await Task.CompletedTask;
            },cancellationToken: destroyCancellationToken);
```

```json
StreamRunAsync -> <b>[POST:200]</b> <color="cyan">https://api.openai.com/v1/threads/thread_yNRqO78fzCl19JYVOmzI0SF1/runs</color> <color="green">Success!</color>
{
  "request": {
    "url": "https://api.openai.com/v1/threads/thread_yNRqO78fzCl19JYVOmzI0SF1/runs",
    "body": {
      "assistant_id": "asst_9iN0RW5a269JNBXNV0VmenYf",
      "model": "gpt-4o",
      "instructions": "Hello! I am using CreateTra....
```
See attached for full log:
[Log.txt](https://github.com/user-attachments/files/16074610/Log.txt)

**Comment by StephenHodgson** [comment]
> Is is possible to not bring back everything and just the delta?

Probably not, as I am using this for something very specific.

That being said there is definitely a bug where the Instructions are being appended with the same text over and over. [Def a bug there](https://github.com/RageAgainstThePixel/com.openai.unity/issues/264).

**Reply by StephenHodgson** [reply]
Ok, oddly enough I'm not actually appending the `Instructions` at all. Seems to be a bug with OpenAI where the Instructions field is getting duplicated somehow by the API itself. Not really sure what is going on there.

**Reply by StephenHodgson** [reply]
@JBlakeMCAA I'm not entirely sure how but from your log file it seems the `Instructions` field grew, and this is the raw response from the API. Sorry this isn't a bug I can fix since it isn't caused by this library.

I ran a few tests and was not able to reproduce the instructions being modified by the API.

If you can get me a solid repro test I'll look into it more.

**Reply by JBlakeMCAA** [reply]
@StephenHodgson  Thank you for looking into this. I have cleaned everything up and ran it again, I've attached the C# code, log and debug results.
If you run the code a few times you'll see a freeze for more than 20 seconds  or greater, could you just double check with the attached that I didn't miss anything?
Log:
[Editor.log](https://github.com/user-attachments/files/16076254/Editor.log)
Debug:
[Unity DebugOutput.txt](https://github.com/user-attachments/files/16076256/Unity.DebugOutput.txt)
C#:
[StreamMessage.txt](https://github.com/user-attachments/files/16076257/StreamMessage.txt)

**Reply by StephenHodgson** [reply]
I would remove and disable the logging as that will definitely add to the overhead of the response time.

**Reply by ElisMia** [reply]
maybe it's some bug from open ai. i also noticed much longer annswers especially when tools are used (Code interpretter), and it got slower in the last monthes, maybe it's some open ai issue... I was checking forums and there is some people complain about it, especially when work with files. 

I noticed that same api that was used before now works slower and sometimes get canceled, because requst takes more then 30 sec.

**Reply by JBlakeMCAA** [reply]
I ended up writing my own and get 6 seconds roundtrip. 
I found:
**Ignore all of the objects coming back in the stream**, except for event: **thread.message.delta** and **event: thread.message.in_progress.** 
I use the message.in_progress to get the messageID and RunID once I have those I ignore it.
Most all of the other response events included the files and instructions and this causes a ton of time serialization them.
Why OpenAI Keeps sending the same things over and over again for each and every stream is ridiculous. They should at least allow setting what you want returned.

**Reply by ElisMia** [reply]
> Why OpenAI Keeps sending the same things over and over again for each and every stream is ridiculous. They should at least allow setting what you want returned.

good to know thanks for info)

**Comment by JBlakeMCAA** [comment]
- If you run the test Unity will freeze up for 20 seconds and then continue if the instructions are greater than 10k, 

- If you have a short instructions it will randomly freeze at random times, even after reading the first few message delta's, 

- If we put a delay of longer then 1 second between streams it will double some of the message delta events.

@StephenHodgson  I've added you as collaborator to the test project, and added a button with code to test:"RAPOpenAI\Assets\Samples\OpenAI\8.1.1\Chat".

It looks to be a semaphore 'wait for signal' or 'wait for command from main thread' issue that is hanging the process.

![Capture](https://github.com/RageAgainstThePixel/com.openai.unity/assets/68558288/77738986-2c8b-408b-a95f-ecfe15e7ec65)
